k8s/plugin/src/resources/upgrade.rs

```diff
@@ -0,0 +1,328 @@
+                if crb.iter().count() == 0 {
+                    let ns: Option<String> = Some(ns.to_string());
+                    let role_binding = objects::upgrade_operator_cluster_role_binding(ns);
+                    match uo
+                        .upgrade_cluster_role_binding
+                        .create(&pp.clone(), &role_binding)
+                        .await
+                    {
+                        Ok(_) => {
+                            println!("cluster role binding created");
+                        }
+                        Err(e) => {
+                            println!("Failed in creating cluster role binding {:?}", e);
+                            tokio::time::sleep(Duration::from_secs(1)).await;
+                            std::process::exit(1);
+                        }
+                    }
+                } else {
+                    println!("cluster role binding already present!");
+                }
+
+                // Create Deployment
+                let crb = uo.get_deployment().await;
+                if crb.iter().count() == 0 {
+                    let ns: Option<String> = Some(ns.to_string());
+                    let upgrade_deploy =
+                        objects::upgrade_operator_deployment(ns, UPGRADE_IMAGE.to_string());
+                    match uo.upgrade_deployment.create(&pp, &upgrade_deploy).await {
+                        Ok(_) => {
+                            println!("deployment created");
+                        }
+                        Err(e) => {
+                            println!("Failed in creating deployment {:?}", e);
+                            tokio::time::sleep(Duration::from_secs(1)).await;
+                            std::process::exit(1);
+                        }
+                    }
+                } else {
+                    println!("deployment already present!");
+                }
+
+                // Create Service
+                let crb = uo.get_service().await;
+                if crb.iter().count() == 0 {
+                    let ns: Option<String> = Some(ns.to_string());
+                    let upgrade_service = objects::upgrade_operator_service(ns);
+                    match uo.upgrade_service.create(&pp, &upgrade_service).await {
+                        Ok(_) => {
+                            println!("service created");
+                        }
+                        Err(e) => {
+                            println!("Failed in creating service {:?}", e);
+                            tokio::time::sleep(Duration::from_secs(1)).await;
+                            std::process::exit(1);
+                        }
+                    }
+                } else {
+                    println!("service already present!");
+                }
+            }
+            Err(e) => println!("Failed to install. Error {}", e),
+        };
+    }
+
+    /// Uninstall the upgrade resources
+    pub async fn uninstall(ns: &str) {
```

Comment:
Uninstall order should be reverse of the install order.
This way, the operator can act on SIGTERM/SIGINT signals and make kube API requests. If the clusterrole or clusterrolebinding is deleted, then the operator code loses all privileges.

Delete Deployment
wait for pod with app=operator-upgrade to go away (if this fails after a timeout, then we can add a `--force` flag which will add deleteoptions to implement --grace-period=0 --force pod delete)
Once the Pod is no longer there....
Delete Service
Delete Clusterrolebinding
Delete Clusterrole
Delete ServiceAccount

Reply:
`Force` flag can be taken in another PR  later. For now when i issue a delete deployment , i must be assured that its deleted, then only i can delete all others.